File: algorithms/sorting/heap_sort/heap_sort_with_array.cpp
#include <iostream>

using namespace std;

#define REP(i, n) for(int i = 0; i < n; i++)

class Heap
{
  private:
    void swap(int *a, int *b)
    {
      int temp = *a;
      *a = *b;
      *b = temp;
    }

    void max_heapify(int *a, int heapsize, int i)
    {
      int l = (2*i + 1),
          r = (2*i + 2),
          largest = i;

      if (l < heapsize && a[l] > a[largest])
        largest = l;

      if (r < heapsize && a[r] > a[largest])
        largest = r;

      if (largest != i) {
        swap(&a[largest], &a[i]);
        max_heapify(a, heapsize, largest);
      }
    }

    void build_max_heap(int *a, int heapsize)
    {
      for(int i = heapsize/2 - 1; i >= 0; --i)
        max_heapify(a, heapsize, i);
    }

  public:

    Heap(int *a, int size)
    {
      build_max_heap(a, size);
    }

    void sort(int *a, int heapsize)
    {
      int temp;

      for(int i = heapsize - 1; i >= 0; --i) {
        swap(&a[0], &a[i]);

        max_heapify(a, i, 0);
      }
    }
};


void display_array(int *a, int size)
{
  REP(i, size) cout << a[i] << "  ";
  cout << endl;
}

int main()
{
  int *a, size;

  cout << "What size array do you want? ";
  cin >> size;

  a = new int[size];
  cout << "Enter array elements : ";
  REP(i, size) cin >> a[i];

  cout << "Current array : ";
  display_array(a, size);

  Heap arr_heap = Heap(a, size);
  cout << "Array Heap : ";
  display_array(a, size);
  arr_heap.sort(a, size);

  cout << "Sorted array : ";
  display_array(a, size);

  return 0;
}

File: algorithms/sorting/heap_sort/heap_sort_with_array.rb
module Heap

  def max_heapify(heapsize, i)
    l = 2*i + 1
    r = l + 1
    largest = i

    if l < heapsize && self[l] > self[largest]
      largest = l
    end

    if r < heapsize && self[r] > self[largest]
      largest = r
    end

    if largest != i
      swap(i, largest)
      max_heapify(heapsize, largest)
    end
  end

  def build_max_heap
    for i in (0..(size/2 - 1)).to_a.reverse
      max_heapify(size, i)
    end
  end
end

class Array

  include Heap

  def swap(i, j)
    temp = self[i]
    self[i] = self[j]
    self[j] = temp
  end

  public

    def heap_sort!
      build_max_heap
      for i in (0..(size - 1)).to_a.reverse
        swap(0, i)
        max_heapify(i, 0)
      end
    end
end

def interact
  puts "What size array do you want ? "
  size = gets.chomp.to_i

  array = []
  puts "Enter the array elements : "
  size.times { array.push(gets.chomp.to_i) }

  puts "Current array : "
  puts array.inspect

  array.heap_sort!

  puts "Sorted array : "
  puts array.inspect
end

interact